llsd/base.py

```diff
@@ -375,36 +370,41 @@ class LLSDBaseParser(object):
     """
     Utility methods useful for parser subclasses.
     """
-    __slots__ = ['_buffer', '_index', '_decode_buff']
+    __slots__ = ['_error', '_peek_impl', '_getc', '_decode_buff']
+
+    # set True to force StreamUtil implementation even for bytes
+    force_stream = True
 
     def __init__(self):
-        self._buffer = b''
-        self._index = 0
+        self._reset(b'')
         # Scratch space for decoding delimited strings
         self._decode_buff = bytearray(_DECODE_BUFF_ALLOC_SIZE)
 
-    def _error(self, message, offset=0):
-        try:
-            byte = self._buffer[self._index+offset]
-        except IndexError:
-            byte = None
-        raise LLSDParseError("%s at byte %d: %s" % (message, self._index+offset, byte))
-
-    def _peek(self, num=1):
+    def _reset(self, something):
+        if isinstance(something, bytes):
+            util = (StreamUtil(io.BytesIO(something)) if self.force_stream
+                    else BytesUtil(something))
+        elif hasattr(something, 'seek'):
+            util = StreamUtil(something)
+        else:
+            raise LLSDParseError('Input must either be a bytes object '
+                                 'or a seekable binary stream')
+        # Instead of indirecting through 'util' on every access, directly
+        # capture its published methods as methods on this instance, but
+        # prepend an underscore to each method name.
+        for attr in dir(util):
+            if not attr.startswith('_'):
+                setattr(self, '_' + attr, getattr(util, attr))
```

Comment:
Why...? Let's avoid metaprogramming for minor convenience reasons if possible.

Reply:
That was actually intended for performance: avoiding an extra hash lookup. There's code in our parsers that uses the Python idiom of capturing certain object attributes as local variables. I don't often do that myself because I'm not often dealing with performance-critical Python code, and I prefer clarity. That said, the lines above are a nod to the theoretical performance benefit of skipping an extra indirection.

But with the local stream wrappers vanishing, the above becomes moot and will likewise disappear.